services/src/main/java/org/keycloak/authentication/authenticators/browser/AbstractUsernameFormAuthenticator.java

```diff
@@ -48,6 +49,7 @@
 
     public static final String REGISTRATION_FORM_ACTION = "registration_form";
     public static final String ATTEMPTED_USERNAME = "ATTEMPTED_USERNAME";
+    private boolean isUserPasswordForm = false;
```

Comment:
I vote for removing this private variable entirely from this class. Instead of it, there could be possibly some method defined on AbstractUsernameFormAuthenticator like for example:
```
/**
 * @return Error message displayed to the user when he enters invalid data (EG. "Invalid username or password" or "Invalid username"
 */
protected String getDefaultChallengeMessage() {
    return Messages.INVALID_USER;
}
```

This method then can be overriden on classes like UsernameForm and PasswordForm.

The various calls to "challenge" can look like this: 
```
challengeResponse = challenge(context, getDefaultChallengeMessage());
```

I believe this will simplify the PR and there won't be so much changes needed to the method "validateUserAndPassword" . Ideally no changes will be needed in this method :)

Reply:
Thanks for your advice! That would be better approach :))